src/lib/checksum.lua

```diff
@@ -68,6 +68,62 @@ function verify_packet (buf, len)
    return ipsum(buf+headersize, len-headersize, initial) == 0
 end
 
+local
+function prepare_packet_l4( buf, len, csum_start, csum_off)
+
+  local hwbuf =  ffi.cast('uint16_t*', buf)
+
+  local pheader = C.pseudo_header_initial(buf, len)
+  if band(pheader, 0xFFFF0000) == 0 then
+    hwbuf[(csum_start+csum_off)/2] = C.htons(band(pheader, 0x0000FFFF))
+  else
+    csum_start, csum_off = nil, nil
+  end
+
+  return csum_start, csum_off
+end
+
+function prepare_packet4( buf, len)
+
+  local hwbuf =  ffi.cast('uint16_t*', buf)
+  local proto = buf[9];
+
+  local csum_start = lshift(band(buf[0], 0x0F),2)
+  local csum_off
+
+  -- Update the IPv4 checksum (use in-place pseudoheader, by setting it to 0)
+  hwbuf[5] = 0;
+  hwbuf[5] = C.htons(C.cksum_generic(buf, csum_start, 0));
```

Comment:
This would appear to call cksum_generic on all transmitted packets.  AFAIU we should not be doing this -- it's the app's responsibility.  We don't do this on intel cards and it seems there is no reason to do it on virtio interfaces.  Besides that, this calls the generic checksum routine, not the avx2 optimized one; at the least we should be using lib.checksum.ipsum instead.